crates/wasmtime/src/component/linker.rs

```diff
@@ -230,6 +231,59 @@ impl<T> LinkerInstance<'_, T> {
         self.insert(name, Definition::Func(func.into_host_func()))
     }
 
+    /// Define a new host-provided function using dynamic types.
+    ///
+    /// `name` must refer to a function type import in `component`.  If and when
+    /// that import is invoked by the component, the specified `func` will be
+    /// called, which must return a `Val` which is an instance of the result
+    /// type of the import.
+    pub fn func_new<F: Fn(&[Val]) -> Result<Val> + Send + Sync + 'static>(
+        &mut self,
+        component: &Component,
+        name: &str,
+        func: F,
+    ) -> Result<()> {
+        self.func_new_with_context(component, name, move |_, args| func(args))
+    }
+
+    /// Define a new host-provided function using dynamic types.
+    ///
+    /// This is the same as [`LinkerInstance::func_new`] except that it accepts
+    /// a function which accepts a `StoreContextMut` as the first parameter.
+    pub fn func_new_with_context<
+        F: Fn(StoreContextMut<'_, T>, &[Val]) -> Result<Val> + Send + Sync + 'static,
+    >(
+        &mut self,
+        component: &Component,
+        name: &str,
+        func: F,
+    ) -> Result<()> {
+        for (import_name, ty) in component.env_component().import_types.values() {
+            if name == import_name {
+                if let TypeDef::ComponentFunc(index) = ty {
+                    let types = component.types();
+                    let ty = &types[*index];
+                    let name = self.strings.intern(name);
+                    return self.insert(
+                        name,
+                        Definition::Func(HostFunc::new_dynamic(
+                            func,
+                            ty.params
+                                .iter()
+                                .map(|(_, ty)| Type::from(ty, types))
+                                .collect(),
+                            Type::from(&ty.result, types),
```

Comment:
This looks like it's basically copying the results of `ty`, but could `ty` be used as-is? The pairing of `index: TypeFuncIndex` and `&Arc<Types>` I think should be enough to type-check everything

Reply:
Ok that seems reasonable, this can be iterated on later if necessary.